### PR TITLE
Add User-Agent header

### PR DESCRIPTION
**Pull request by thewilkybarkid**
/cc @giorgiosironi

### CODE REVIEW

File: spec/ApiClient/BlogClientSpec.php [hunk 1]
```diff
@@ -37,7 +38,7 @@ public function it_gets_an_article()
     public function it_lists_articles()
     {
         $request = new Request('GET', 'blog-articles?page=1&per-page=20&order=desc&subject[]=cell-biology',
-            ['X-Foo' => 'bar', 'Accept' => 'application/vnd.elife.blog-article-list+json; version=2']);
+            ['X-Foo' => 'bar', 'Accept' => 'application/vnd.elife.blog-article-list+json; version=2', 'User-Agent' => 'eLifeApiClient/'.Version::get()]);
```
Comment (giorgiosironi) [inline spec/ApiClient/BlogClientSpec.php]: looks like a missing abstraction, did we have a `stubRequest()` method around?

File: src/ApiClient.php [hunk 2]
```diff
@@ -54,6 +55,13 @@ public function __construct(HttpClient $httpClient, array $headers = [])
 
         $request = new Request('PUT', $uri, $headers, $content);
 
+        return $this->send($request);
+    }
+
+    private function send(RequestInterface $request) : PromiseInterface
+    {
+        $request = $request->withHeader('User-Agent', trim(sprintf('eLifeApiClient/%s %s', Version::get(), $request->getHeader('User-Agent')[0] ?? '')));
```
Comment (giorgiosironi) [inline src/ApiClient.php]: why it is necessary to add the value also here? Who uses this trait?

File: src/Version.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            self::create();
+        }
+
+        return self::$version;
+    }
+
+    private static function create()
+    {
+        $originalVersion = Versions::getVersion('elife/api-client');
+        list($version) = explode('@', $originalVersion);
```
Comment (giorgiosironi) [inline src/Version.php]: what happens if the value is not what we expect? `list()` is silently throwing away part of it?

File: src/Version.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        if (false !== strpos($version, 'dev')) {
+            $version = $originalVersion;
+        }
```
Comment (giorgiosironi) [inline src/Version.php]: if `$version` is a sha, can we retain just the first 7 characters?

File: src/ApiClient.php (hunk 2 above)
Comment (thewilkybarkid) [inline src/ApiClient.php]: Each of the clients. I've done it in two places to separate adding the `eLifeApiClient` details from the `GuzzleHttp` details.

File: spec/ApiClient/BlogClientSpec.php (hunk 1 above)
Comment (thewilkybarkid) [inline spec/ApiClient/BlogClientSpec.php]: Separately I've been tempted to remove PHPSpec and replace these with plain unit tests. We can refactor later.

**Comment by thewilkybarkid** [comment]
`UserAgentPrependingHttpClient` has been abstracted so that the SDK, Journal etc can use it quickly. I don't like `Version` as that will need to duplicated everywhere... unless it's abstracted to a separate library, which feels like overkill (`ocramius/package-versions` will _always_ include the full Git hash, even if it's a tagged released).

File: src/Version.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+final class Version
+{
+    private static $version;
+
+    private function __construct()
+    {
+    }
+
+    public static function get() : string
```
Comment (giorgiosironi) [inline src/Version.php]: a smoke test over this (regexp over its return value) could be useful to better pin down what it can produce

**Pull request merged**